### PR TITLE
Added function `pretty` for pretty printing JSON

### PR DESCRIPTION
**Pull request by larsewi**
Signed-off-by: larsewi <lars.erik.wik@northern.tech>

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thanks for submitting a PR! Maybe @vpodzime can review this?

File: cfbs/pretty_json.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+import json
+from typing import OrderedDict
+
+
+def pretty(s):
+    MAX_LEN = 80
+    INDENT_SIZE = 2
+
+    def _force_wrap(parent):
```
Comment (olehermanse) [inline cfbs/pretty_json.py:9]: `should_wrap` or `should_force_wrap` like a question.

File: cfbs/pretty_json.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        if len(parent) < 2:
+            return False
+        for child in parent:
+            if not isinstance(child, (tuple, list, dict)):
+                return False
+            if len(child) < 2:
+                return False
```
Comment (olehermanse) [inline cfbs/pretty_json.py:16]: Count number of children with 2 or more elements. If count is more than 1, wrap.

File: cfbs/pretty_json.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                    first = False
+                else:
+                    buf += ", "
+                if not isinstance(key, str):
+                    raise ValueError("Illegal key type '" + type(key).__name__ + "'")
+                buf += '"' + key + '": '
+                buf += _encode(value, indent, cursor + len(buf))
+            buf += " }"
+
+            if indent + cursor + len(buf) <= MAX_LEN:
```
Comment (olehermanse) [inline cfbs/pretty_json.py:73]:
```suggestion
            assert "\n" not in buf
            if indent + cursor + len(buf) <= MAX_LEN:
```

File: cfbs/pretty_json.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                    break
+                if first:
+                    first = False
+                else:
+                    buf += ", "
+                if not isinstance(key, str):
+                    raise ValueError("Illegal key type '" + type(key).__name__ + "'")
+                buf += '"' + key + '": '
+                buf += _encode(value, indent, cursor + len(buf))
+            buf += " }"
```
Comment (olehermanse) [inline cfbs/pretty_json.py:71]: If this is only for 1 line anyway, could you just use `json.dumps()`?

Comment (olehermanse) [reply cfbs/pretty_json.py:71]: or map or list / dict comprehension.

File: cfbs/pretty_json.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                    break
+                if first:
+                    first = False
+                else:
+                    buf += ", "
+                buf += _encode(value, indent, cursor + len(buf))
+            buf += "]"
+
+            if indent + cursor + len(buf) <= MAX_LEN:
+                return buf
```
Comment (olehermanse) [inline cfbs/pretty_json.py:37]: Unclear whether buf can be multiple lines or not. Please change to json.dumps, or add asserts and map/join.

**Review by olehermanse — CHANGES_REQUESTED** [review]

File: cfbs/pretty.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+import json
+from json import encoder
+from typing import OrderedDict
```
Comment (vpodzime) [inline cfbs/pretty.py:3]: `from collections import OrderedDict`

File: cfbs/pretty.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        assert isinstance(parent, (tuple, list, dict))
+
+        if isinstance(parent, dict):
+            parent = parent.values()
+
+        count = 0
+        for child in parent:
+            if isinstance(child, (tuple, list, dict)):
+                if len(child) >= 2:
+                    count += 1
```
Comment (vpodzime) [inline cfbs/pretty.py:20]: I'd prefer a bit more functional-like approach here, something like:
```
count = len([child for child in parent if isinstance(child, (tuple, list, dict) and len(child) >= 2])
return rount >= 2
```
But YMMV.

File: cfbs/pretty.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        if data is None:
+            return "null"
+        elif data is True:
+            return "true"
+        elif data is False:
+            return "false"
+        elif isinstance(data, int):
+            return int.__repr__(data)
+        elif isinstance(data, float):
+            return float.__repr__(data)
```
Comment (vpodzime) [inline cfbs/pretty.py:85]: You could merge these two cases into
```
elif isinstance(data, (int, float)):
    return repr(data)
```

File: test/test_pretty.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            '{ "This": "line", "is": "equal", "to": 80, "characters": "dont\'t wrap me...." }'
+        )
+        == '{ "This": "line", "is": "equal", "to": 80, "characters": "dont\'t wrap me...." }'
+    )
+    assert (
+        pretty(
+            '{ "This": "line", "is": "greater", "than": 80, "characters": "wrap me ........" }'
+        )
+        == '{\n  "This": "line",\n  "is": "greater",\n  "than": 80,\n  "characters": "wrap me ........"\n}'
+    )
```
Comment (vpodzime) [inline test/test_pretty.py:41]: I'd prefer this multiline cases to be checked like this:
```
expected = """{
  "This": "line",
  "is": "greater",
  ...
}"""
assert(pretty(...) == expected)
```

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise. Please change the unit test to show the structure of the expected format better. The current format makes it very hard to review.

File: cfbs/pretty.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        elif isinstance(data, str):
+            return '"' + data + '"'
+        elif isinstance(data, (list, tuple)):
+            return _encode_list(data, indent, cursor)
+        elif isinstance(data, dict):
+            return _encode_dict(data, indent, cursor)
+        else:
+            raise ValueError("Illegal value type '" + type(data).__name__ + "'")
+
+    data = json.loads(s, object_pairs_hook=OrderedDict)
```
Comment (olehermanse) [inline cfbs/pretty.py:95]: Should be possible to call a pretty function with a dictionary. 9/10 times I want to generate some dictionary, and then prettify it before saving to file or printing to terminal.

Maybe it makes sense to copy what the json module is doing (`load()`, `loads()`, `dump()`, `dumps()`). You could have `pretty()` and `pretty_string()`.

**Review by olehermanse — CHANGES_REQUESTED** [review]

File: cfbs/pretty.py (hunk 7 above)
Comment (larsewi) [reply cfbs/pretty.py:20]: @vpodzime, @olehermanse  how about ?
```Python
count = len(
    filter(
        lambda child: isinstance(child, (tuple, list, dict))
        and len(child) >= 2,
        parent,
    )
)
```

**Review by olehermanse — APPROVED** [review]
LGTM, just a small comment.

**Pull request merged**